tests/cnf/core/network/metallb/tests/bgp-gracefulrestart.go

```diff
@@ -127,6 +144,157 @@ var _ = Describe("BGP Graceful Restart", Ordered, Label(tsparams.LabelGRTestCase
+				verifyMetalLbBFDAndBGPSessionsAreUPOnFrrPod(frrPod, ipv4NodeAddrList)
+
+				By("Checking that BGP GracefulRestart is enabled on the neighbors")
+				verifyGREnabledOnNeighbors(frrPod, ipv4NodeAddrList)
+			})
+
+			AfterEach(func() {
+				By("Applying the MachineConfig to remove the created nft policy")
+				createMCAndWaitforMCPStable(tsparams.DeleteAllRules)
+				err := mco.NewMCBuilder(APIClient, "98-nft").Delete()
+				Expect(err).ToNot(HaveOccurred(), "Failed to get the machineConfig")
+
+				By("Checking that BGP and BFD sessions are established and up")
+				verifyMetalLbBFDAndBGPSessionsAreUPOnFrrPod(frrPod, ipv4NodeAddrList)
+			})
+
+			It("Test Compatibility with BFD", reportxml.ID("77101"), func() {
+				gracefulRestartTestWithBFD("Cluster", addressPool[0], ipv4MetalLbIPList[0], frrPod, ipAddressPool, workerLabelMap)
+			})
+		})
+		Context("multi hop", func() {
+			var (
+				speakerRoutesMap  map[string]string
+				masterClientPodIP = "172.16.0.1"
+				frrMasterIPs      = []string{"172.16.0.253", "172.16.0.254"}
+			)
+			BeforeEach(func() {
+				By("Collecting information before test")
+				frrk8sPods, err := pod.List(APIClient, NetConfig.MlbOperatorNamespace, metav1.ListOptions{
+					LabelSelector: tsparams.FRRK8sDefaultLabel,
+				})
+				Expect(err).ToNot(HaveOccurred(), "Failed to list frrk8s pods")
+
+				speakerRoutesMap, err = buildRoutesMap(frrk8sPods, ipv4MetalLbIPList)
+				Expect(err).ToNot(HaveOccurred(), "Failed to build frrk8s route map")
+
+				By("Creating internal NAD")
+				masterBridgePlugin, err := nad.NewMasterBridgePlugin("internalnad", "br0").
+					WithIPAM(nad.IPAMStatic()).GetMasterPluginConfig()
+				Expect(err).ToNot(HaveOccurred(), "Failed to create master bridge plugin setting")
+				internalNad, err := nad.NewBuilder(APIClient, "internal", tsparams.TestNamespaceName).
+					WithMasterPlugin(masterBridgePlugin).Create()
+				Expect(err).ToNot(HaveOccurred(), "Failed to create internal NAD")
+
+				By("Configuring Local GW mode")
+				setLocalGWMode(true)
+
+				By("Creating FRR pod one on master node")
+				createFrrPodOnMasterNodeAndWaitUntilRunning("frronmaster1",
+					ipv4MetalLbIPList[0], netparam.IPSubnet24, frrMasterIPs[0], internalNad.Definition.Name,
+					masterNodeList[0].Object.Name, addressPool[0], ipv4NodeAddrList[0])
+
+				By("Creating FRR pod two on master node")
+				createFrrPodOnMasterNodeAndWaitUntilRunning("frronmaster2",
+					ipv4MetalLbIPList[1], netparam.IPSubnet24, frrMasterIPs[1], internalNad.Definition.Name,
+					masterNodeList[0].Object.Name, addressPool[0], ipv4NodeAddrList[1])
+
+				By("Creating FRR pod in the test namespace")
+				frrPod = createFrrPod(
+					masterNodeList[0].Object.Name,
+					bfdConfigMap.Object.Name,
+					[]string{},
+					pod.StaticIPAnnotation(internalNad.Definition.Name,
+						[]string{fmt.Sprintf("%s/%s", masterClientPodIP, netparam.IPSubnet24)}))
+
+				// Add static routes from client towards Speaker via router internal IPs
```

Comment:
Is a bit hard to follow the network setup, where is the client, what is the routing table on client, why we need three frrpods, what are the routing table on each on, is there any  diagram/readme?

E.g. https://github.com/metallb/metallb/tree/main/e2etest#bgp-tests-network-topology this is what we have in metallb

Reply:
The multihop topology can be found in Polarion, just above the BFD multihop tests.
https://polarion.engineering.redhat.com/polarion/#/project/OSE/wiki/CNF/CNF%20MetalLB?selection=OCP-47186